<html><head></head><body>
<h2 class="h2" id="ch10"><span epub:type="pagebreak" id="page_325"/><span class="big">10.</span> biology and bioinformatics</h2>&#13;
<p class="noindent">I once came across a beautiful diagram in <em>Science</em><a href="footnote.xhtml#fn29" id="fn_29">*</a> showing the metabolic pathways of one of the smallest bacteria, <em>Mycoplasma pneumoniae</em>. It reminded me of staring at an Apple II schematic when I was less than a decade old. Back then, I knew that the Apple II schematic’s fascinatingly complex mass of lines was a map to the computer in front of me, though I didn’t know quite enough to do anything with that map. But the point was that a map existed, so despite its imposing appearance, it gave me hope that I could unravel such complexities. Biological “schematics” like the one on the next page give me the same hope.</p>&#13;
<div class="image"><span epub:type="pagebreak" id="page_326"/><img alt="image" src="../images/f0326-01.jpg"/></div>&#13;
<p class="cap">Mycoplasma pneumoniae<em>’s metabolic pathway</em></p>&#13;
<div class="image"><img alt="image" src="../images/f0326-02.jpg"/></div>&#13;
<p class="cap"><em>The Apple II schematic from my wall</em></p>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_327"/>The <em>M. pneumoniae</em> diagram isn’t quite as precise as the Apple II schematic, but from 10,000 feet, they feel similar in complexity and detail. The metabolic diagram is detailed enough for me to trace a path from glucose to ethanol, and the Apple II schematic is detailed enough for me to trace a path from the CPU to the speaker. And just as a biologist wouldn’t make much of a box with 74LS74 attached to it, an electrical engineer wouldn’t make much of a box with ADH inside it. (A 74LS74 contains two instances of a synchronous electronic storage device, and ADH is alcohol dehydrogenase, an enzyme coded by gene <em>MPN564</em> that can turn acetaldehyde into ethanol.)</p>&#13;
<p class="indent">Furthering the computer analogy, though, the <em>Science</em> article’s authors also included a list that read like a BOM for <em>M. pneumoniae</em> in their supplemental material. The pentagonal boxes in the diagram are <em>enzymes</em>, proteins that catalyze specific chemical reactions. Each enzyme is listed with a functional description along with its gene sequence, which is equivalent to source code.</p>&#13;
<p class="indent">At the very end of that list, I saw a table of uncharacterized genes. If you’ve done a bit of reverse engineering, you’ve probably made similar tables for parts or function calls in an electronic system. They’re the first place I go for fresh clues when I get stuck. I find it heartening to see biologists and hackers applying similar techniques to reverse engineering complex systems.</p>&#13;
<h3 class="h3" id="ch00lev1sec185"><strong>COMPARING H1N1 TO A COMPUTER VIRUS</strong></h3>&#13;
<p class="noindenta">The comparison of biological systems to computer systems doesn’t stop at the metabolic level. I once read a fascinating article in <em>Nature</em><a href="footnote.xhtml#fn30" id="fn_30">*</a> that compared the pathogenic components <span epub:type="pagebreak" id="page_328"/>of the <em>novel H1N1 virus</em> (better known as <em>swine flu</em>) to those of other flu strains, and that article got me thinking about how digital and organic viruses compare. For example, how big is an organic virus relative to a digital one? To put the question another way, how many bits does it take to kill a human, or at least make one quite sick? In exploring this idea, I found it helpful to draw a few analogies between the digital and organic worlds.</p>&#13;
<h3 class="h3" id="ch00lev1sec186"><strong>DNA and RNA as Bits</strong></h3>&#13;
<p class="noindenta">When the H1N1 pandemic broke out in 2009, the virus was comprehensively sequenced and logged in the National Center for Biotechnology Information’s (NCBI) Influenza Virus Resource database, and the data collected there is amazing. I love the specificity of the records. For example, the entire sequence of an instance of influenza known as A/Italy/49/2009(H1N1) isolated from the nose of a 26-year-old female <em>Homo sapiens</em> returning from the United States to Italy is on the NCBI website. Here are the first 120 bits of the DNA sequence:</p>&#13;
<pre>atgaaggcaa tactagtagt tctgctatat acatttgcaa ccgcaaatgc agacacatta</pre>&#13;
<p class="indent">With 120 bits total, each symbol (A, T, G, or C) represents 2 bits of information. In genes, this can be alternatively represented as an amino acid sequence, where every three DNA symbols are a <em>codon</em> corresponding to one amino acid. Long chains of <em>amino acids</em> fold into complex structures called <em>proteins</em> that give structure and function to a cell, and chains of amino acids too short to be a complete protein are often called <em>peptides</em>. Using a translation lookup table that biologists call the standard genetic code, I converted the previous sequence into the following peptide: <code>MKAILVVLLYTFATANADTL</code>.</p>&#13;
<p class="indent">In this sequence, each symbol represents an amino acid, which is the equivalent of six bits or three DNA bases per amino acid. There are 20 amino acids in the canonical codon <span epub:type="pagebreak" id="page_329"/>table, and each letter corresponds to a different amino acid. M is methionine, K is lysine, A is alanine, and so on.</p>&#13;
<p class="indent">Now, consider RNA, which passes information from DNA on how to synthesize proteins to the rest of the cell. As with DNA, each base in RNA specifies one of four possible symbols (in this case, A, U, G, or C), so a single base corresponds to two bits of information. DNA and RNA are information-equivalent on a one-to-one mapping. Think of DNA as a program stored on disk and RNA as the same program loaded into RAM. When DNA is loaded, protein synthesis instructions are transcribed into RNA, but all T bases are replaced with U bases.</p>&#13;
<p class="indent">Proteins, then, are the output of running an RNA program. Proteins are synthesized according to the instructions in RNA on a three-to-one mapping. You can think of proteins like pixels in a frame buffer, as follows:</p>&#13;
<p class="bull">• A complete protein is like an image on the screen.</p>&#13;
<p class="bull">• Each amino acid on a protein is like a pixel.</p>&#13;
<p class="bull">• Each pixel has a depth of six bits, due to the three-to-one mapping of a medium that stores two bits per base.</p>&#13;
<p class="bull">• Finally, each pixel goes through a color palette (the codon translation table) to transform the raw data into a final rendered color. Unlike a computer frame buffer, however, different biological proteins vary in amino acid count (analogous to a pixel count).</p>&#13;
<p class="indentd">To ground this in a specific example, imagine that six bits stored as ATG on your hard drive (DNA) are loaded into RAM (RNA) as AUG because T is transcribed as U when going from DNA to RNA. When the RNA program in RAM is executed, AUG is translated to a pixel (amino acid) of color M, or methionine, which is the biological “start” codon—that is, the first instruction in every valid RNA program.</p>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_330"/>As a shorthand, since DNA and RNA are one-to-one equivalent, bioinformaticists represent gene sequences in DNA format, even if the biological mechanism is in RNA format. The influenza virus has an RNA architecture, rather than DNA, and the 120 bits of DNA I showed earlier correspond to an RNA subroutine in influenza. That subroutine codes for the HA gene, which produces an H1 variety of the hemagglutinin protein. This is the <em>H1</em> in the H1N1 designation of swine flu.</p>&#13;
<h4 class="h4" id="ch00lev1sec187"><strong>Organisms Have Unique Access Ports</strong></h4>&#13;
<p class="noindenta">Given that background information, if you think of organisms as computers with IP addresses, each functional group of cells in the organism listens to the environment through its own active port. As port 25 maps specifically to SMTP services on a computer, port H1 maps specifically to the windpipe region on a human. Interestingly, the same port H1 maps to the intestinal tract on a bird. Thus, the same H1N1 virus will attack the respiratory system of a human and the gut of a bird. In contrast, H5—the variety of hemagglutinin protein found in H5N1, the deadly avian flu—specifies the port for your inner lungs. As a result, H5N1 is much deadlier than H1N1 because it attacks your inner lung tissue, causing severe pneumonia. H1N1 is less deadly because it attacks a more benign port that just makes you blow your nose a lot and cough up loogies.</p>&#13;
<div class="note">&#13;
<p class="notet"><strong>NOTE</strong></p>&#13;
<p class="notep"><em>Researchers are still discovering more about the H5 port. The</em> Nature <em>article I read indicated that perhaps certain human mutants have lungs that don’t listen on the H5 port. People whose lungs ignore the H5 port would have a better chance of surviving an avian flu infection, while those that open port H5 on the lungs have no chance to survive (make your time ... all your base pairs are belong to H5N1).<a href="footnote.xhtml#fn31" id="fn_31">*</a></em></p>&#13;
</div>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_331"/>Knowing a virus is deadly, you can figure out how many bits it takes to kill a human (or at least make one quite sick) by calculating the number of bits in the viral genome. The question, then, is how many bits are in this instance of H1N1? The raw number of bits, by my count, is 26,022; the number of actual coding bits is approximately 25,054. I say “approximately” because in some places, the virus does the equivalent of self-modifying code to create two proteins out of a single gene. It’s hard to say what counts as code and what counts as an incidental, nonexecuting NOP sled required for the self-modified code.</p>&#13;
<p class="indent">That means it takes about 25Kb or 3.2KB of data to code for a virus that has a nontrivial chance of killing a human. This is more efficient than a computer virus like MyDoom, which comes in around 22KB. Knowing that I could be killed by 3.2KB of genetic data is humbling. Then again, with roughly 800MB of data in my genome, there’s bound to be an exploit or two.</p>&#13;
<h4 class="h4" id="ch00lev1sec188"><strong>Hacking Swine Flu</strong></h4>&#13;
<p class="noindenta">One interesting consequence of reading this <em>Nature</em> article and having access to the virus sequence is that in theory, I now know how to modify the virus sequence to make it deadlier. For instance, the <em>Nature</em> article notes that variants of the PB2 influenza gene with glutamic acid at position 627 in the sequence have a <em>low pathogenicity</em>, meaning they aren’t very deadly. However, PB2 variants with lysine at the same position increase the likelihood of mortality.</p>&#13;
<p class="indent">Let’s see the sequence of PB2 for H1N1. Going back to the NCBI database, I found the following amino acid sequences around position 627:</p>&#13;
<pre><span class="ash">601 QQMRDVLGTFDTVQIIKLLP</span><br/><span class="ash">621 FAAAPP</span><span class="codestrong">E</span><span class="ash">QSRMQFSSLTVNV</span><br/><span class="ash">641 RGSGLRILVRGNSPVFNYNK</span></pre>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_332"/>The numbers to the left indicate the position of the first symbol in each line of the sequence; I’ll follow that convention for the rest of this discussion. Check the line labeled 621, and note the E in position 627. E is the symbol for glutamic acid. Thankfully, H1N1 seems to be a less-deadly version of influenza; perhaps this is why fewer people died from contracting H1N1 than the media might have led you to believe.</p>&#13;
<p class="indent">Now, let’s reverse this back to the DNA code:</p>&#13;
<pre><span class="ash">621   F   A   A   A   P   P</span>   <span class="codestrong">E</span>   <span class="ash">Q   S   R</span><br/><span class="ash">1861 ttt gct gct gct cca cca</span> <span class="codestrong">gaa</span> <span class="ash">cag agt agg</span></pre>&#13;
<p class="indent">Notice the GAA codes for E. To modify this genome to be deadlier, you’d simply need to replace GAA with one of the codes for lysine (K). Lysine can have a code of either AAA or AAG. Thus, a deadlier variant of H1N1 would have a coding sequence like this:</p>&#13;
<pre><span class="ash">621   F   A   A   A   P   P</span>   <span class="codestrong">K</span>   <span class="ash">Q   S   R</span><br/><span class="ash">1861 ttt gct gct gct cca cca</span> <span class="codestrong">aaa</span> <span class="ash">cag agt agg</span><br/>                             ^ changed</pre>&#13;
<p class="indent">So, a single base-pair change—simply flipping two bits— might be all you’d need to turn the H1N1 swine flu virus into a deadlier variant. Theoretically, I could apply a series of well-known biological procedures to synthesize this strain and actually implement the hack. As a first step, I could go to a DNA synthesis website and order the modified sequence to get my deadly little project going for just over $1,000. Some of those companies have screening procedures to protect against DNA sequences that could be used to implement biohazardous products, but even if they happened to screen for HA variants, there are well-known protocols for site-directed mutagenesis that could possibly be used to modify a single base of RNA from material extracted from normal H1N1.</p>&#13;
<h4 class="h4" id="ch00lev1sec189"><span epub:type="pagebreak" id="page_333"/><strong>Adaptable Influenza</strong></h4>&#13;
<p class="noindenta">Of course, I have to give influenza some credit. It packs a deadly punch in 3.2KB, and despite scientists’ best efforts, we haven’t eradicated it. Could influenza do hacks like the one I just described on its own already?</p>&#13;
<p class="indent">The short answer is yes.</p>&#13;
<p class="indent">In fact, the influenza virus evolved to allow for these adaptations. Normally, when DNA is copied, an error-checking protein runs over the copied genome to verify that no mistakes were made. This keeps the error rate quite low. But remember, the influenza virus uses an RNA architecture. It therefore needs a different mechanism from DNA for copying.</p>&#13;
<p class="indent">Inside its protein capsule, the influenza virus packs code for a protein complex called <em>RNA-dependent RNA polymerase</em>, which is a tiny machine for copying RNA off of RNA templates. Normally, RNA is only generated by transcribing DNA, not by copying an existing piece of RNA, so this mechanism is essential for the replication of RNA-based influenza. Significantly, RNA-dependent RNA polymerase omits an error-checking protein that would prevent mutations. The result is that influenza makes about one error per 10,000 base pairs that get copied. The influenza genome is about 13,000 base pairs long, so on average, every copy of an influenza virus has one random mutation.</p>&#13;
<p class="indent">Some of these mutations make no difference; others render the virus harmless; and quite possibly, some render the virus much more dangerous. Since viruses are replicated and distributed in astronomical quantities, the chance that this little hack could end up occurring naturally is in fact quite high. I think this is part of the reason health officials were so worried about H1N1: people had no resistance to it, and even though it wasn’t as deadly as it could have been, the strain was probably just a couple of mutations away from being a much bigger health problem.</p>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_334"/>There is one other important subtlety to the RNA architecture of the influenza virus, aside from its high mutation rate: the virus’s genetic information is stored as eight separate, relatively short, snippets of RNA. In many other viruses and simple organisms, genetic information is instead stored as a single unbroken strand.</p>&#13;
<p class="indent">To understand why that’s important, consider what happens when a host is infected by two types of the influenza virus at the same time. If the genes were stored as a single piece of DNA or RNA, there would be little opportunity for the genes between the two types to shuffle. But because influenza stores its genes as eight separate snippets, those genes mix freely inside the infected cell and are randomly shuffled into virus packets as they emerge. If you’re unlucky enough to get two types of flu at once, the result is a potentially novel strain of flu, as RNA strands are copied, mixed, picked out of the metaphorical hat, and then packed into virus particles. This process is elegant in that the same mechanism allows for mixing of an arbitrary number of strains in a single host. If you can infect a cell with three or four types of influenza at once, the result is an even wilder variation of flu particles.</p>&#13;
<p class="indent">This mechanism is part of the reason novel H1N1 is called a <em>triple-reassortant</em> virus. Through a series of dual infections or perhaps a single calamitous infection of multiple flu varieties, novel H1N1 acquired a mix of RNA snippets that gave it high transmission rates and made it something humans weren’t innately immune to. That’s the perfect storm for a pandemic.</p>&#13;
<p class="indent">If there were a computer analogy to this RNA-shuffling model, it would be a virus that distributes itself in the form of unlinked object code files plus a small helper program that, upon infecting a host, relinks its files in a random order before copying and redistributing itself. It would also search for similar viruses that may already be infecting that computer and on occasion link in object code with matching function <span epub:type="pagebreak" id="page_335"/>templates from the other viruses. This rearrangement and novel relinking of the code itself would foil classes of antivirus software that search for virus signatures based on fixed code patterns. It would also proliferate a diverse set of viruses in the wild, with less predictable properties.</p>&#13;
<p class="indent">The influenza virus’s multilevel adaptation mechanism is remarkable. The virus has both a slowly evolving point mutation mechanism and a mechanism for drastically altering its properties in a single generation through gene-level mixing with other viruses. It doesn’t work quite like sex, but the result is probably just as good, if not better. It’s also remarkable that these two important properties of the virus arise as a consequence of using RNA instead of DNA as the genetic storage medium.</p>&#13;
<h3 class="h3" id="ch00lev1sec190"><strong>A Silver Lining</strong></h3>&#13;
<p class="noindenta">Since there are so many variants of flu, no vaccine can target all types of the virus, but the H1N1 story does have a silver lining. Apparently, a patient who contracted swine flu during the pandemic created a novel antibody with the remarkable ability to confer immunity to all 16 subtypes of influenza A. A group of researchers sifted through the patient’s white blood cells and managed to isolate four B cells that contained the code to produce this antibody. They cloned the cells and produced antibodies, facilitating further research into a potential vaccine that could confer broad protection against the flu.</p>&#13;
<p class="indent">I found this really interesting at a gut level because it gives me hope that if a killer virus did wipe out most of humanity, maybe a small group of people would survive it.</p>&#13;
<h3 class="h3" id="ch00lev1sec191"><strong>REVERSE ENGINEERING SUPERBUGS</strong></h3>&#13;
<p class="noindenta">In 2011, a “superbug” strain of <em>E. coli</em> (a species of bacteria with subtypes that can cause food poisoning) called EHEC O104:H4 broke out in Europe. When I found out that scientists <span epub:type="pagebreak" id="page_336"/>at BGI, located in Shenzhen, had released the entire sequence of O104:H4 freely online for anyone to examine, I got very curious about the situation. I couldn’t help but wonder exactly what tools bioinformaticists use to analyze DNA sequences. Manually inspecting the relatively simple sequences of the influenza virus is one thing, but there must be computational tools to help make sense of more complicated organisms like <em>E. coli</em>.</p>&#13;
<p class="indent">Fortunately, my perlfriend (<code>s/perl/girl/</code>) is also a noted bioinformaticist. She took some time out of her busy schedule to show me some tools of the trade. It turns out most of the tools for analyzing DNA are freely available online. Since DNA is just sequences of A’s, T’s, G’s, and C’s, the standard data interchange format is plain old ASCII text, which means you can do a lot of analysis using command-line tools like <code>grep</code>, <code>sed</code>, and <code>awk</code>.</p>&#13;
<h4 class="h4" id="ch00lev1sec192"><strong>The O104:H4 DNA Sequence</strong></h4>&#13;
<p class="noindenta">The raw sequence data BGI provided was a set of oversampled subsequences that we needed to assemble by matching up overlapping regions. Stitching subsequences together is a bit like composing a large picture from small photos taken at random. With enough sampling, you’ll eventually create a mostly complete picture, but the image will still have ambiguities, particularly in areas with regular patterns.</p>&#13;
<p class="indent">The genome of O104:H4 was provided as a list of over 500,000 short DNA samples. The assembly process stitched the short DNA samples together into 513 contiguous fragments of DNA (known as <em>contigs</em>), with a total genome length of 5.3 million base pairs. An organism like <em>E. coli</em> has just one big loop of DNA, so there were 513 spots where limitations in the sequencing technology (or just bad luck) missed an unknown number of base pairs, preventing us from knowing the entire, unbroken sequence. Notably, a typical, non-superbug strain of <span epub:type="pagebreak" id="page_337"/><em>E. coli</em> has around 4.6 million base pairs, so O104:H4 is probably at least 15 percent longer. Likewise, this strain would take more time to replicate than a non-drug-resistant strain. Take a look at contig 34 of the assembly:</p>&#13;
<pre><span class="ash">AAATGGTATTCCTGTTCACGATACTATTGCCAGAGTTGTATCCTGTATCAGTCCTGC</span><br/><span class="ash">AAAATTTCATGAGTGCTTTATTAACTGGATGCGTGACTGCCATTCTTCAGATGATAA</span><br/><span class="ash">AGACGTCATTGCAATTGATGGAAAAACGCTCCGGCACTCTTATGACAAGAGTCGCCG</span><br/><span class="ash">CAGGGGAGCGATTCATGTCATTAGTGCGTTCTCAACAATGCACAGTCTGGTCATCGG</span><br/><span class="ash">ACAGATCAAGACGGATGAGAAATCTAATGA</span><span class="codestrong">GATTACA</span><span class="ash">GCTATCCCAGAACTTCTTAA</span><br/><span class="ash">CATGCTGGATATTAAAGGAAAAATCATCACAACTGATGCGATGGGTTGCCAGAAAGA</span><br/><span class="ash">TATTGCAGAGAAGATACAAAAACAGGGAGGTGATTATTTATTCGCGGTAAAAGGAAA</span><br/><span class="ash">CCAGGGGCGGCTAAATAAAGCCTTTGAGGAAAAATTTCCGCTGAAAGAATTAAATAA</span><br/><span class="ash">TCCAGAGCATGACAGTTACGCAATTAGTGAAAAGAGTCACGGCAGAGAAGAAA</span></pre>&#13;
<p class="indent">I could have picked any contig, and it probably would have made about as much sense to you as this block of letters. Aside from making gratuitous pop culture references (the word <em>GATTACA</em> occurs 252 times in the genome of O104:H4), the raw DNA sequence isn’t very insightful. It’s a bit like staring at binary machine code. To analyze the data, you need to “decompile” the “methods” contained within the code.</p>&#13;
<p class="indent">In this case, we were searching for DNA sequences that code for <em>proteins.</em> As I mentioned earlier, proteins are complex, often interwoven chains of molecules consisting of small building blocks known as amino acids. Cells get things done using proteins: some proteins turn sugar into energy, others use that energy to move around or change the cell’s shape, and still others are responsible for copying and repairing the cell.</p>&#13;
<p class="indent">Fortunately, protein sequences are highly conserved in DNA. Nature tends to reuse protein structures, with few modifications, between organisms. Thus, a function that has been determined through a biological experiment, even on another species, can often be correlated with a sequence of DNA. For instance, one common experiment for determining the function of a sequence is to cut a piece of DNA out of a cell and observe what happens to the cell; the loss of function <span epub:type="pagebreak" id="page_338"/>resulting from the missing DNA is often indicative of the protein’s role in the cell.</p>&#13;
<p class="indent">Biologists have amassed decades of research on what certain proteins do into huge databases. Thus, to figure out what a chunk of DNA means, you can do a fuzzy pattern match between your DNA of interest and the database of known proteins.</p>&#13;
<h4 class="h4" id="ch00lev1sec193"><strong>Reversing Tools for Biology</strong></h4>&#13;
<p class="noindenta">I needed two tools to reverse engineer DNA: a protein database and a piece of software called BLASTX. Both are free to download online.</p>&#13;
<h5 class="h5" id="ch00lev1sec194">THE UNIPROT DATABASE</h5>&#13;
<p class="noindenta">I downloaded a list of known proteins from the Universal Protein Resource, or UniProt (<em><a href="http://www.uniprot.org/">http://www.uniprot.org/</a></em>). In 2011, a search of the database for “drug resistance” restricted to <em>E. coli</em> organisms yielded a list of 1,378 proteins that scientists have identified over the years as parts of the <em>E. coli</em> bacteria’s drug-resistance machinery. Every year, new discoveries are added to the database.</p>&#13;
<p class="indent">Here’s a snippet from the database that describes a protein that gives O104:H4 resistance to a drug you may recognize:</p>&#13;
<pre>&gt;sp|P0AD65|<span class="codestrong">PBP2_ECOLI Penicillin-binding protein 2</span><br/>OS=Escherichia coli (strain K12) GN=mrdA PE=3 SV=1<br/><br/>MKLQNSFRDYTAESALFVRRALVAFLGILLLTGVLIANLYNLQIVRFTDYQTRSNENRIK<br/>LVPIAPSRGIIYDRNGIPLALNRTIYQIEMMPEKVDNVQQTLDALRSVVDLTDDDIAAFR<br/>KERARSHRFTSIPVKTNLTEVQVARFAVNQYRFPGVEVKGYKRRYYPYGSALTHVIGYVS<br/>KINDKDVERLNNDGKLANYAATHDIGKLGIERYYEDVLHGQTGYEEVEVNNRGRVIRQLK<br/>EVPPQAGHDIYLTLDLKLQQYIETLLAGSRAAVVVTDPRTGGVLALVSTPSYDPNLFVDG<br/>ISSKDYSALLNDPNTPLVNRATQGVYPPASTVKPYVAVSALSAGVITRNTTLFDPGWWQL<br/>PGSEKRYRDWKKWGHGRLNVTRSLEESADTFFYQVAYDMGIDRLSEWMGKFGYGHYTGID<br/>LAEERSGNMPTREWKQKRFKKPWYQGDTIPVGIGQGYWTATPIQMSKALMILINDGIVKV<br/>PHLLMSTAEDGKQVPWVQPHEPPVGDIHSGYWELAKDGMYGVANRPNGTAHKYFASAPYK<br/>IAAKSGTAQVFGLKANETYNAHKIAERLRDHKLMTAFAPYNNPQVAVAMILENGGAGPAV<br/>GTLMRQILDHIMLGDNNTDLPAENPAVAAAEDH</pre>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_339"/>PBP2_ECOLI<a href="footnote.xhtml#fn32" id="fn_32">*</a> is linked to penicillin resistance and is a mutated gene that determines the shape of the bacteria. It seems this resistant variant adapted to operate despite the presence of penicillin; bacteria with nonresistant forms of the gene are unable to form properly shaped cell walls in the presence of penicillin, and are killed by the drug. Other genes might cause more active countermeasures, like pumping an antibiotic out of the cell or modifying the antibiotic to be less toxic to the cell. Browsing the UniProt database gives you a feel for the huge variety of genes available in nature that can make bacteria resistant to drugs.</p>&#13;
<h5 class="h5" id="ch00lev1sec195">THE DECOMPILER</h5>&#13;
<p class="noindenta">Next, I needed the actual decompiler. That’s where BLASTX (eventually updated to BLAST+) came in. BLASTX is a variant of BLAST, which stands for <em>Basic Local Alignment Search Tool</em>. First, I had this analysis program compute all possible translations of the <em>E. coli</em> DNA to protein sequences. Translating DNA results in six possible protein sequences: DNA can be read forward and backward (known as 5′→3′ and 3′→5′), and each direction has three possible frame positions. Then, I had the program check for patterns among the resulting amino acid sequences that matched the database of sequences known to provide drug resistance. (I could have also checked for other types of patterns, by typing something different into the database query.) The result was a sorted list of each known drug resistance protein, along with the region of the <em>E. coli</em> genome that best matches the protein.</p>&#13;
<p class="indent">The following is the BLASTX output for the penicillin example.</p>&#13;
<pre><span epub:type="pagebreak" id="page_340"/># BLASTX 2.2.24 [Aug-08-2010]<br/><br/># Query: 43 87880<br/># Database: uniprot-drug-resistance-AND-organism-coli.fasta<br/># Fields: Query id, Subject id, % identity, alignment length,<br/>mismatches, gap openings, q. start, q. end, s. start, s. end,<br/>e-value, bit score<br/>43 sp|P0AD65|<span class="codestrong">PBP2_ECOLI 100.00</span> 632 0 0 29076 30971 1 632 0.0 1281<br/>43 sp|P0AD68|FTSI_ECOLI 25.08 650 458 21 29064 30926 6 574 2e-33 142<br/>43 sp|P60752|MSBA_ECOLI 32.80 186 120 6 12144 12686 378 558 6e-17 87.0<br/>43 sp|P60752|MSBA_ECOLI 27.78 216 148 5 77054 77677 361 566 8e-14 76.6<br/>43 sp|P77265|MDLA_ECOLI 27.98 193 133 6 12141 12701 370 555 2e-10 65.5<br/><br/>--<span class="codeitalic">snip</span>--</pre>&#13;
<p class="indent">The Fields line describes what each column in the table shows. In the % identity column, you can see that the gene for PBP2_ECOLI has a 100 percent match inside the genome of O104:H4.</p>&#13;
<h5 class="h5" id="ch00lev1sec195a"><b>Answering Biological Questions with UNIX Shell Scripts</b></h5>&#13;
<p class="noindent">With this list, I could answer some interesting questions, like “How many of the known drug resistance genes are inside O104:H4?” Here’s the one-liner program that my perlfriend wrote to answer that particular question:</p>&#13;
<pre>cat uniprot_search_m9 | awk '{if ($3 == 100) { print;}}' | \<br/>  cut -f2 |grep -v ^# | cut -f1 -d"_" | cut -f3 -d"|" | \<br/>  sort | uniq | wc -l</pre>&#13;
<p class="indent">The output from that script told us that 1,138 genes in O104:H4 were a 100 percent match against the database of 1,378 genes that can confer drug resistance. When we loosened the criteria to also list 99 percent matches, allowing for one or two mutations per gene, the list expanded to 1,224 out of 1,378. The “superbug” O104:H4 earned its title, having acquired roughly 90 percent of the known resistance genes!</p>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_341"/>I also wanted to answer the inverse question: which drug-resistance genes are most definitely not in O104:H4? By looking at the resistance genes missing from a superbug, we might be able to gather clues as to which treatments could be effective against the bug.</p>&#13;
<p class="indent">To rule out a drug-resistance gene, we crafted another search that would reveal which resistance genes in the database had less than a 70 percent match against the sequence of O104:H4. The 70 percent threshold was just an arbitrary number I picked; there’s probably a rigorous standard that scientists and clinicians use.</p>&#13;
<p class="indent">Here is the list, as it appeared in my terminal:</p>&#13;
<pre><span class="ash">A0SKI3 A2I604 A3RLX9 A3RLY0 A3RLY1 A5H8A5 B0FMU1 B1A3K9 B1LGD9<br/>B3HN85 B3HN86 B3HP88B5AG18 B6ECG5 B7MM15 B7MUI1 B7NQ58 B7NQ59<br/>B7TR24</span> <span class="codestrong">BLR</span> <span class="ash">CML D2I9F6 D5D1U9 D5D1Z3 D5KLY6 D6JAN9 D7XST0 D7Z7R4<br/>D7Z7W9 D7ZDQ3 D7ZDQ4 D8BAY2 D8BEX8 D8BEX9 DYR21 DYR22 DYR23<br/>E0QC79 E0QC80 E0QE33 E0QF09 E0QF10 E0QYN4 E1J2I1 E1S2P1 E1S2P2<br/>E1S382 E3PYR0 E3UI84 E3XPK9 E3XPQ2 E4P490 E5ZP70 E6A4R5 E6A4R6<br/>E6ASX0 E6AT17 E6B2K3 E6BS59 E7JQV0 E7JQZ4 E7U5T3 E9U1P2 E9UGM7<br/>E9VGQ2 E9VX03 E9Y7L7 O85667 Q05172 Q08JA7 Q0PH37 Q0T948 Q0T949<br/>Q0TI28 Q1R2Q2 Q1R2Q3 Q3HNE8 Q4HG53 Q4HG54 Q4HGV8 Q4HGV9 Q4HH67<br/>Q4U1X2 Q4U1X5 Q50JE7 Q51348 Q56QZ5 Q56QZ8 Q5DUC3 Q5UNL3 Q6PMN4<br/>Q6RGG1 Q6RGG2 Q75WM3 Q79CI3 Q79D79 Q79DQ2 Q79DX9 Q79IE6 Q79JG0<br/>Q7BNC7 Q83TT7 Q83ZP7 Q8G9W6 Q8G9W7 Q8GJ08 Q8VNN1 Q93MZ2 Q99399<br/>Q9F0D9 Q9F0S4 Q9F7C0 Q9F8W2 Q9L798</span></pre>&#13;
<p class="indent">You can plug any of these protein codes into the UniProt database and find out more about them. For example, BLR is beta-lactamase, an enzyme that causes resistance to beta-lactam antibiotics. UniProt describes it like this:</p>&#13;
<p class="termp12">Has an effect on the susceptibility to a number of antibiotics involved in peptidoglycan biosynthesis. Acts with beta lactams, D-cycloserine and bacitracin. Has no effect on the susceptibility to tetracycline, chloramphenicol, gentamicin, fosfomycin, vacomycin or quinolones. Might enhance drug exit by being part of multisubunit efflux pump. Might also be involved in cell wall biosynthesis.</p>&#13;
<p class="indent">Unfortunately, a cursory inspection revealed that most functions that O104:H4 lacked were just small, poorly understood <span epub:type="pagebreak" id="page_342"/>fragments of machines involved in drug resistance. As a result, there was no clear candidate for a superbug killer in its genome.</p>&#13;
<h4 class="h4" id="ch00lev1sec196"><strong>More Questions Than Answers</strong></h4>&#13;
<p class="noindenta">The good news is that anyone can access the tools to analyze genomes, and some tools, such as grep, awk, and sed, are already familiar to computer engineers. The bad news is that while we can ask questions about the genome with these tools, we’re still left with more questions than answers. For example, antibiotic resistance sounds like a good thing for the survival of bacteria, so why don’t all bacteria have it? And how do bacteria go about acquiring (or losing) such genes?</p>&#13;
<p class="indent">The rise of antibiotic-resistant superbugs is a product of our love of antibiotics. As DNA in <em>E. coli</em> copies at a rate of about a dozen base pairs per second, shedding even a single unused gene can lend a meaningful advantage in an exponential growth race; after all, an <em>E. coli</em> population can double every 20 minutes in optimal conditions. As a result, there is selective pressure to shed genes that aren’t necessary for survival. The genome of O104:H4 is 15 percent longer than that of a typical <em>E. coli</em> strain, which means that after seven generations, a typical <em>E. coli</em> strain would have twice the population of O104:H4. Within half a day under optimal, antibiotic-free growth conditions, a strain of <em>E. coli</em> unburdened with antibiotic resistance genes would have over 20 times the population of O104:H4. Thus, a bacterium that hangs on to its antibiotic resistance genes is like a sprinter wearing a bulletproof vest to a race. Likewise, one of the greatest natural threats to superbugs is a lean, fast-replicating common bug that can edge out the superbug by sheer numbers alone.</p>&#13;
<p class="indent">However, bacteriocidal and bacteriostatic antibiotics kill off or prohibit growth of nonresistant bugs, respectively, leaving only the resistant bugs to grow unhindered. Over time and with exposure to several types of antibiotics, it stands to reason <span epub:type="pagebreak" id="page_343"/>that the resistant bug population would continue to selectively breed for multiple resistance genes, creating a superbug.</p>&#13;
<p class="indent">Still, I find it astonishing that resistant bugs seem to develop resistance genes so quickly. We’re taught that evolution is a slow process, so it seems remarkable that bacteria can serendipitously evolve a suite of antibiotic resistance genes totaling hundreds of thousands of base pairs. New genes do in fact take a very long time to spontaneously arise (there are very few clearly documented cases of this, such as the Long-Term Evolution Experiment by Richard Lenski). Instead, most resistance genes are acquired from the environment through <em>horizontal gene transfer</em>.</p>&#13;
<p class="indent">Our environment is teeming with DNA fragments. The GitHub of biology is all around us, from the dirt to the sea to the air we breathe. Some DNA fragments code for useful traits; some are just junk. When a bacterium is under stress (like it is when exposed to antibiotics), it may start to take up random DNA fragments from the environment and manufacture proteins based off the code. If it’s going to die anyway, it might as well, right? Most of the time, the incorporated DNA fragments are not helpful, but if one lucky bacterium picks up the necessary resistance gene from the environment, it can rapidly outcompete others in an antibiotic-laden environment.</p>&#13;
<p class="indent">Thus, while nonresistant strains of a bug will rapidly outnumber antibiotic-resistant strains, the tiny remaining population of resistant bugs (or perhaps even their lifeless bodies floating about in the environment) form a reservoir of genetic material that can be drafted in times of stress. And since the genetic code is interoperable across all species, resistance genes can even be acquired from unrelated organisms.</p>&#13;
<p class="indent">Discovering that the functions O104:H4 lacked were poorly understood was an interesting lesson in itself. Fiction popularizes the notion that knowing a DNA sequence is the same as knowing what diseases or traits an organism may have. But <span epub:type="pagebreak" id="page_344"/>even though we know the sequences and general properties of many proteins, it’s much harder to link proteins to a specific disease or trait. At some point, someone has to get their hands dirty and do biological experiments involving actual organisms to assign biological significance to a given protein family.</p>&#13;
<p class="indent">Pop culture references to DNA analysis are glibly unaware of this missing link in the process, which leads to overinflated expectations for genetic analysis, particularly in its utility for diagnosing and curing human disease and applications in eugenics. Let’s take a closer look at some of those myths.</p>&#13;
<h3 class="h3" id="ch00lev1sec197"><strong>MYTHBUSTING PERSONALIZED GENOMICS</strong></h3>&#13;
<p class="noindenta">We’re definitely living in The Future in a lot of ways. For instance, we have electric cars! But Hollywood reels from the ’60s and ’70s also predicted that I’d be using a flying car to get around town by now, not just an electric car on the ground. Of course, automotive technology isn’t the only victim of Hollywood hype.</p>&#13;
<p class="indent">The potential impact of personalized genomics is greatly overstated in movies like <em>GATTACA</em>, which create a myth that your genome is like a crystal ball, and somehow your fate is predestined by your genetic programming. The perlfriend I mentioned earlier coauthored a paper in <em>Nature</em><a href="footnote.xhtml#fn33" id="fn_33">*</a> examining 23andMe’s direct-to-consumer (DTC) personal genomics offerings. Let’s have a look at her paper, and let the mythbusting begin!</p>&#13;
<h4 class="h4" id="ch00lev1sec198"><strong>Myth: Having Your Genome Read Is Like Hex-Dumping the ROM of Your Computer</strong></h4>&#13;
<p class="noindent">An inexpensive technique to look at parts of the genome is called <em>genotyping</em>. Here, a selective diff is done between your genome and a reference human genome; in other words, your <span epub:type="pagebreak" id="page_345"/>genome is simply sampled in potentially interesting spots for single-point mutations called <em>single nucleotide polymorphisms</em> (SNPs, pronounced “snips”). The concept of genotyping naturally leads to two questions. First, how do you decide which SNPs are interesting enough to sample? And second, how do you know the reference genome is an accurate comparison point? This sets up two more busted myths.</p>&#13;
<h4 class="h4" id="ch00lev1sec199"><strong>Myth: We Know Which Mutations Predict Disease</strong></h4>&#13;
<p class="noindenta">Some mutations in the human genome simply correlate with disease; they are not proven to be predictive or causal. In truth, we really don’t understand why many genetic diseases happen. For poorly understood diseases, all we can say is that people who have a particular disease tend to have a certain pattern of SNPs. It’s important not to confuse causality with correlation.</p>&#13;
<p class="indent">Thus, while scientists can make predictions about diseases based on SNPs, most of those predictions are correlative, not causative (and weakly correlative, at that). As a result, a genotype should not be considered a crystal ball for predicting your disease future. Rather, it’s closer to a Rorschach blot that you have to squint and stare at for a while before you can say what it means. For instance, in the paper my perlfriend wrote, she found that companies often didn’t match up on their predictions for disease risk because they interpreted mutation meanings differently.</p>&#13;
<h4 class="h4" id="ch00lev1sec200"><strong>Myth: The Reference Genome Is an Accurate Reference</strong></h4>&#13;
<p class="noindenta">The word <em>reference</em> in <em>reference genome</em> should tip you off on a problem: it implies there are “reference people.” Ultimately, just a handful of individuals were sequenced to create today’s reference genome, and most of them are of European ancestry. As time goes on and more full-sequence genetic data is collected, the reference genome will be merged and massaged <span epub:type="pagebreak" id="page_346"/>to present a more accurate picture of the overall human race, but for now, it’s important to remember that a genotype study is a diff against a source repository of questionable universal validity.</p>&#13;
<p class="indent">For example, some SNPs have different frequencies in different populations. The base A might dominate in a European population, but at that same position in an African population, the base G could dominate. It’s also important to remember that the reference genome has an aggregate error rate of about 1 error in 10,000 base pairs, although to be fair, the process of discovering a disease variant usually cleans up any errors in the reference genome for the relevant sequence regions.</p>&#13;
<p class="indent">It will be decades before we have a full understanding of what all the sequences in the human genome mean, and even then, they may not be truly predictive of disease risk or anything else about our health. Here lies perhaps the most important message, and a point I can’t stress enough: in most situations, environment has more to do with who you are, what you will become, and what diseases you will have than your genes do. Any upside to personal genomics won’t be due to crystal-ball predictions, but rather to the fact that knowing about their own genetic predispositions may encourage more people to make lifestyle changes that will help them stay healthy. If there’s one thing I’ve learned from dating a preeminent bioinformaticist, it’s that no matter your genetic makeup, most common diseases can be prevented or delayed with proper diet and exercise.</p>&#13;
<h3 class="h3" id="ch00lev1sec201"><strong>PATCHING A GENOME</strong></h3>&#13;
<p class="noindenta">So far in this chapter, I’ve given examples of sequencing and analyzing genomes. That’s more or less the equivalent of being able to dump a program executable and analyze it in IDA. Oftentimes, after you analyze an executable, you’ll want to <span epub:type="pagebreak" id="page_347"/>patch it to do something new. Patching software is relatively straightforward and reliable: just fire up a hex editor and change the file. In the worst case, you might have to use a focused ion beam (FIB) to modify the individual wires of a mask ROM inside a chip.</p>&#13;
<p class="indent">But historically, the ability to patch a genome has been severely limited. Information in cells is stored at the molecular level, and changing a specific portion of a gene can be a painstaking process. Just as vacuum tubes and transistors came before the integrated circuit, zinc finger nucleases (ZFNs) and transcription activator-like effector nucleases (TALENs) enabled gene editing, but with significant caveats in efficiency, performance, and ultimately, cost. In 2012, the integrated circuit of gene editing was introduced: the CRISPR/Cas<a href="footnote.xhtml#fn34" id="fn_34">*</a> system.</p>&#13;
<h3 class="h3" id="ch00lev1sec202"><strong>CRISPRs in Bacteria</strong></h3>&#13;
<p class="noindenta">CRISPR, short for <em>clustered regularly interspaced short palindromic repeat</em>, describes a particular RNA structure, while Cas are proteins that associate with CRISPRs. CRISPRs are, as far as biologists know, common only in bacteria and archaea (for example, fungi), and they’re part of a devilishly clever system for immunity in simple organisms. Like humans, bacteria have immune systems that can be programmed through exposure to pathogens. When bacteria encounter a viral invader, they have proteins that can snip out short sequences of the viral DNA and archive the sequences as spacers in a CRISPR.</p>&#13;
<p class="indent">Labs that failed for months to edit a gene using TALENs switched to CRISPR/Cas and succeeded on the first try. They succeeded so quickly because the process just involves designing a short snippet of RNA that’s inserted into a CRISPR, a simple exercise that can be done entirely on a computer or, I <span epub:type="pagebreak" id="page_348"/>daresay, by hand. The RNA snippet itself can be fabricated in about a week for less than $50 using one of several service providers, replacing a significant amount of wet lab complexity with an informatics exercise.</p>&#13;
<p class="indent">Each CRISPR region is tagged by a leader sequence, immediately followed by the CRISPR proper. A CRISPR itself consists of a guide RNA (gRNA) or “spacer” sequence delimited by a well-defined DNA <em>direct repeat</em> sequence that is palindromic.</p>&#13;
<div class="note">&#13;
<p class="notet"><strong>NOTE</strong></p>&#13;
<p class="notep"><em>The term</em> spacer <em>is used when discussing an immune system, while</em> guide RNA <em>is used when discussing genome editing. Calling a region of interest a spacer is confusing, but misnomers can happen with reverse engineering. I can’t blame scientists for first noticing a pattern of repeating delimiters and calling the stuff between the delimiters “spacers.” After all, physicists got the current flow convention backward and stuck with it. Who are we to judge?</em></p>&#13;
</div>&#13;
<p class="indent">Palindromic typically means that a string is equivalent when simply reversed, like the word <em>racecar</em>. When biologists say a sequence is “palindromic,” they mean the sequence is equivalent when first complemented (A→T, T→A, G→C, C→G) and then reversed. For instance, GAATTC is considered biologically palindromic, even though it is not lexically palindromic.</p>&#13;
<p class="indent">The CRISPR/Cas system was described shortly after the demise of Chumby, and at the time, I was interning at Dr. Swaine Chen’s infectious diseases laboratory at the Genome Institute of Singapore. Among other things, I studied various strains of <em>E. coli</em> that induce urinary tract infection, under the guidance of Lu Ting Liow. While assisting an investigation into portions of phage virus DNA that found its way into <em>E. coli</em>, I was asked to write a script to identify palindromic and repeating sequences of DNA in the <em>E. coli</em> genome. My script showed that the genome was littered with the sequences; I figured the code had a bug and didn’t think much of the result. <span epub:type="pagebreak" id="page_349"/>But perhaps some of the direct repeats I saw were portions of a CRISPR.</p>&#13;
<p class="indent">Let’s look at a CRISPR from a strain of <em>E. coli</em> now. This is the CRISPR direct repeat sequence for <em>E. coli</em> O104:H4:</p>&#13;
<pre><span class="ash">GAGT</span><span class="codestrong">TCCCCGC</span><span class="ash">GCCA</span><span class="codestrong">GCGGGGA</span><span class="ash">TAAACCG</span></pre>&#13;
<p class="indent">The bolded base pairs are the palindromic regions. When this DNA sequence is translated into RNA (so that T→U), the palindromic region can pair with itself, forming a hairpin or stem loop, as shown here.</p>&#13;
<div class="image"><img alt="image" src="../images/f0349-01.jpg"/></div>&#13;
<p class="cap"><em>A stem loop</em></p>&#13;
<p class="indent">This shape hints at the significance of the repeated palindromic structures in a CRISPR: when translated into RNA, the sequence can fold onto itself, forming a <em>secondary structure</em>. It’s important to remember that genes are not just lines of code; they are physical molecules whose overall shape significantly impacts their function. Biologists use a four-tier system for describing the physical structure that molecules like DNA, RNA, and proteins can take based on their source code. Primary structure is simply the sequence of monomers <span epub:type="pagebreak" id="page_350"/>(bases or amino acids). <em>Secondary structure</em> refers to physical shapes that arise from the localized interactions of monomers, due to physical properties such as the spacing and number of hydrogen bonds between molecules, or the affinity of certain monomers for water. In RNA and DNA, that means structures like hairpin loops; in proteins, it means structures like spirals and sheets. Tertiary structure refers to the complex 3D shape of a molecule that arises from long-distance interactions between potentially remote portions of the primary sequence. Tertiary structure is particularly applicable to proteins, as some amino acids, such as cysteine, can cross-link with each other over longer distances. Quaternary structure refers to structures formed from the interaction of multiple molecules. A Cas9/RNA complex is an example of a quaternary structure. The final, chemically active and targeted molecule arises only when a Cas9 protein is merged with a gRNA, and the stem loop secondary structure of the gRNA is necessary for Cas9 to recognize it.</p>&#13;
<h4 class="h4" id="ch00lev1sec203"><strong>Determining Where to Cut a Gene</strong></h4>&#13;
<p class="noindenta">RNA derived from a CRISPR region through transcription is incorporated into a protein complex with other Cas proteins. Specific Cas proteins (such as Cas9) use the RNA as a search-and-destroy template: the Cas9/RNA complexes float around the cell, and when they find a DNA sequence that matches the RNA template, they selectively cut the DNA at the template site, effectively neutralizing the intruding virus. But you may have noticed a recursion problem: the Cas9/RNA complex should also cut up the CRISPR region in the host organism’s genome, as that region also has the target pattern. This would effectively destroy the CRISPR region for future use.</p>&#13;
<p class="indent">To avoid destroying the CRISPR region, the Cas9/RNA complex targets the template DNA plus a short, defined three-to-five base pair sequence called a <em>proto-space adjacent motif</em> <span epub:type="pagebreak" id="page_351"/>(PAM). For example, the PAM for a popular Cas9 protein from <em>S. pyogenes</em> is [AGTC]GG when written in regular expression format; biologists use a different convention, NGG, to say the same thing. As long as the CRISPR archive doesn’t include the PAM sequence, it won’t be cut up by the complex.</p>&#13;
<p class="indent">The PAM requirement means there are some limitations on where you can cut a gene. It’s a bit like targeting only hex strings that end in 0xC3 or searching for return-oriented programming (ROP) gadgets. Just as hackers searching for ROP gadgets look for short sequences of instructions that end in a RET opcode, bioinformaticists have to search for short sequences of DNA to edit that end in a PAM.</p>&#13;
<p class="indent">Despite these limitations, CRISPR/Cas has proven to be a versatile and reliable gene-editing tool. It has been adapted to both cut genes and paste in new sequences. Making a precise cut at an arbitrary location in DNA is the hardest step of inserting new DNA. But in conjunction with well-studied techniques like non-homologous end joining (NHEJ) or homology-directed repair (HDR), CRISPR/Cas can be used to insert modifications into a gene.</p>&#13;
<h4 class="h4" id="ch00lev1sec204"><strong>Implications for Engineering Humans</strong></h4>&#13;
<p class="noindenta">Even though CRISPR/Cas is a naturally occurring system found in bacteria and fungi, the universal genetic code means the system is binary-compatible with all species, including humans. Before this system was discovered, genes were largely read-only, especially in living organisms. CRISPR/Cas gives us a much more reliable and efficient tool to patch and repair genes, without necessarily disrupting the viability of the host organism. Biologists have managed to pack the necessary DNA for a CRISPR/Cas exploit into viruses, enabling them to sneak these gene-editing tools through the cell walls of live, complex organisms like mice, plants, and humans. The structure of a CRISPR also allows scientists to perform multiple edits in a <span epub:type="pagebreak" id="page_352"/>single experiment, expanding the experimental and therapeutic versatility of the technique.</p>&#13;
<p class="indent">This technology has already been validated on human cells, even human embryos, and the implications are simply mind-boggling. Regardless of ethical standards set by the scientific and legal communities in your country of residence, I think the promise of custom-designed children, free of genetic diseases that once plagued parents, is too strong a temptation. Even if most countries banned such a practice, I feel it’s inevitable that someone, somewhere, perhaps funded by a wealthy billionaire unable to have viable children of their own, will start tinkering with custom-engineered humans. If the results are positive, it will likely change the course of humanity more profoundly than Moore’s law. And that’s if a mechanism called <em>gene drive</em> doesn’t get there first.</p>&#13;
<h4 class="h4" id="ch00lev1sec205"><strong>Hacking Evolution with Gene Drive</strong></h4>&#13;
<p class="noindenta">Gene drive rewrites the rules of sexual reproduction and, consequently, evolution in a way previously unseen in nature. You might know that you have two copies of every gene: one from your mother and one from your father. Each copy is an <em>allele</em>. If the alleles match, you’re said to be <em>homozygous</em> for that gene. If the alleles are different, you’re <em>heterozygous</em> for it. Normally, which allele a child gets from each parent is a coin toss, and the fitness of a child in a given environment is the primary deciding factor for passing a set of alleles on to a new generation.</p>&#13;
<p class="indent">Gene drive eliminates this coin toss. Environmental selection is short-circuited, allowing genes with potentially negative side effects to propagate rapidly in a population. This exploit is made possible by outfitting the desired allele with a CRISPR/Cas-assisted gene-editing mechanism that targets and converts a heterozygous allele into a homozygous allele. For example, if a mother has a gene outfitted with a CRISPR/Cas-assisted <span epub:type="pagebreak" id="page_353"/>gene drive mechanism, it doesn’t matter what the father’s genes are. Inside the child, the mother’s copy will express the CRISPR/Cas editing mechanisms, seeking out the father’s copy and editing it to be the same as the mother’s.</p>&#13;
<p class="indent">In terms of disruptive power, if CRISPR/Cas is the rm command, then gene drive is like calling <code>rm -r *</code> instead.</p>&#13;
<p class="indent">This has a profound effect on natural selection. Forget survival of the fittest; changes no longer have to strictly benefit an organism’s fitness to spread through the population. Furthermore, gene-driven changes can sweep through a natural population at an exponential rate (much faster than typical mutations) because they don’t rely on coin tosses and natural selection to amplify a mutation.</p>&#13;
<p class="indent">On the upside, gene drive could be used to force good changes into the world, like malaria-free mosquitoes. On the downside, this new mechanism, previously unseen in nature, could wreck havoc on evolution and the ecosystem. Although our changes could be well engineered and well intentioned, nature likes to shake things up through mutations, spontaneous rearrangements, and horizontal gene transfer. If a gene-driven organism were to pick up extra genes in the payload region, the outcome could be unpredictable.</p>&#13;
<p class="indent">For instance, malaria-free mosquitoes would benefit humans, but mosquitoes also play a large role in the Earth’s ecosystem as a food source for fish and birds. If modified mosquitoes failed to thrive and occupy their ecological niche, there could be a domino effect that hurts other species. This could all happen on a timescale so short that we may not be able to reverse it if we tried. Furthermore, organisms like mosquitoes don’t recognize geopolitical boundaries. Thus, banning gene drive in most of the world doesn’t make anyone safe from its potential consequences. If just one well-engineered organism makes it into the wild, everyone has to deal with it.</p>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_354"/>Perhaps it’s no mistake that CRISPR/Cas has been found only in bacteria and archaea—organisms that are known to reproduce asexually. Perhaps the ability to short-circuit the fitness requirement in sexual reproduction rapidly degrades the overall fitness of any germ line carrying a CRISPR/Cas mutation so that the line goes extinct before it can take over a population. After all, any accidental genes or spontaneous mutation that finds its way into a CRISPR/Cas payload would also sweep through the population as quickly as the initial drive.</p>&#13;
<p class="indent">The question, then, is how long does it take for this degradation and extinction to happen? The example of eradicating malaria vectors would have a very different outcome if the modified mosquitoes went extinct within a few years versus several millennia.</p>&#13;
<h3 class="h3" id="ch00lev1sec206"><strong>CLOSING THOUGHTS</strong></h3>&#13;
<p class="noindenta">Clearly, there are a lot of unanswered questions on the frontier of biological engineering, and it’s all happening right now. Whether good or bad, the outcome of today’s experiments will probably affect humanity as profoundly as Moore’s law and the internet. Electronic technology reshaped the way we think and communicate, and biotech will reshape our bodies and our environment. The big difference is that in biotech, we haven’t developed the ability to do backups, but we are developing technology with the potential power of the <code>rm -r *</code> command.</p>&#13;
<p class="indent">Personally, I’m optimistic; I think these technologies can and will be used to improve our lives. But for that to happen, we need society to understand the issues at stake and have a vigorous and open debate. Even if these biological techniques have scary implications for our health and safety, failing to disclose and discuss vulnerabilities just invites zero-days. And who wants to wake up one morning infected with crippling malware and no viable patch?</p>&#13;
<p class="indent"><span epub:type="pagebreak" id="page_355"/>Hardware breakthroughs have changed our lives as we know it, but Moore’s law is slowing down, and DNA sequencing has outpaced it. Who knows what new world will be created by advancements in biotech? And just as society benefits from the responsible disclosure and sharing of vulnerabilities and exploits, engaging in scientific discourse is more constructive than attempting to censor it. Perhaps the experience and perspectives gained in maturing the hardware industry over the past 50 years from pocket calculators into pocket supercomputers can help guide biotech to a similarly positive outcome.<span epub:type="pagebreak" id="page_356"/></p>&#13;
</body></html>